torch/fx/experimental/recording.py

```diff
@@ -0,0 +1,478 @@
+        return len(self.tensor_size)
+
+    @staticmethod
+    def from_fake(fake) -> "FakeTensorMeta":
+        return FakeTensorMeta(fake.size(), fake.stride(), fake.storage_offset())
+
+
+# [Note: ShapeEnv State Equality]
+# ===============================
+#
+# What is considered ShapeEnv state?
+# ----------------------------------
+# We consider to be the state of a ShapeEnv instance everything that
+# is not in the inline tuple inside remove_nonstate_variables function.
+# That is: the fields within ShapeEnv that modify the flow of execution
+# of the program.
+#
+# So, for example: the replacements field might influence on how an
+# expression is simplified. That, in turn, may result in a guard being
+# statically known (i.e. not added).
+#
+# On the other hand, var_to_stack serves only changes what is printed
+# in the screen, i.e. used only for debugging purposes. Therefore, we
+# should not consider it when comparing states.
+#
+# What to do on NotEqualError?
+# ----------------------------
+# Here are a few possible causes for getting a NotEqualError raised:
+#
+#   1. New field that does not belong in the ShapeEnv state.
+#      For example: log field of type ShapeEnvLoggerAdapter. Different
+#      ShapeEnv instances will always have different ShapeEnvLoggerAdapter
+#      instances, i.e. equality comparison would fail.
+#      Solution: add it to the inlined tuple inside remove_nonstate_variables
+#      function inside check_equal method.
+#
+#   2. New field that is not directly comparable across instances.
+#      For example: guards field of type List[ShapeGuard]. More specifically,
+#      the ShapeGuard type holds an expression and a stack information
+#      for debugging purposes. When replaying the even on a new ShapeEnv
+#      instance, the stack would be different, which would trigger this error.
+#      Solution: add a special case to the map_value function inside
+#      check_equal function.
+#
+#   3. Mutation of ShapeEnv on some not recorded function.
+#      If a mutation of the state of ShapeEnv happens inside a function
+#      that is not recorded (or that no caller in the stack is recorded),
+#      then, the replayed ShapeEnv won't catch that.
+#      Solution: decorate the function with record_shape_env_event.
+
+
+# Checks whether the state of two ShapeEnv are equal w.r.t. the guards
+# returned by ShapeEnv.produce_guards.
+def shape_env_check_state_equal(env1, env2):
+    # ShapeEnv fields that are not relevant for the outcome of
+    # ShapeEnv.produce_guards call:
+    #   - Debugging variables
+    #   - Translation validation related variables
+    #   - Events recording related variables
+    state_variable_names = (
+        "counter",
+        "log",
+        "var_to_stack",
+        "fx_node_cache",
+        "graph",
+        "validator",
+        "check_recorded_events",
+        "should_record_events",
+        "is_recording",
+        "tracked_fakes",
+        "events",
```

Comment:
Need to have some sort of way of making sure when new fields are added, they get added here

Reply:
Ah, it's actually the opposite. The name of this variable should've been `non_state_variable_names`. Adding the name of the variable here, means that such a variable is not part of what is considered the state. 

In case someone adds a field to `ShapeEnv` that shouldn't be part of the state, the test will probably call it out.